File: api/src/main/java/com/sniper/music/api/LastFMClient.java
package com.sniper.music.api;

import android.support.annotation.NonNull;

import okhttp3.OkHttpClient;
import retrofit2.CallAdapter;
import retrofit2.Converter;
import retrofit2.Retrofit;

public class LastFMClient implements RetrofitClient {

    public static final String NAME = "LastFMClient";
    public static final String GSON_CONVERTER_FACTORY = "GSONConverterFactory";
    public static final String RX_CALL_ADAPTER_FACTORY = "RXLastFMCallAdapterFactory";

    @NonNull
    private final Retrofit restAdapter;

    public LastFMClient(@NonNull Converter.Factory converterFactory,
                        @NonNull CallAdapter.Factory callAdapterFactory,
                        @NonNull OkHttpClient okHttpClient,
                        @NonNull ApiConfiguration config) {

        restAdapter = new Retrofit.Builder()
                .baseUrl(config.getBaseURL().toExternalForm())
                .addConverterFactory(converterFactory)
                .addCallAdapterFactory(callAdapterFactory)
                .client(okHttpClient)
                .validateEagerly(true)
                .build();
    }

    @Override
    public <T> T api(Class<T> service) {
        return restAdapter.create(service);
    }
}

File: api/src/main/java/com/sniper/music/api/LastFMClient.kt
package com.sniper.music.api

import okhttp3.OkHttpClient
import retrofit2.CallAdapter
import retrofit2.Converter
import retrofit2.Retrofit

class LastFMClient(converterFactory: Converter.Factory,
                   callAdapterFactory: CallAdapter.Factory,
                   okHttpClient: OkHttpClient,
                   config: ApiConfiguration) : RetrofitClient {

    private val restAdapter: Retrofit = Retrofit.Builder()
            .baseUrl(config.baseURL.toExternalForm())
            .addConverterFactory(converterFactory)
            .addCallAdapterFactory(callAdapterFactory)
            .client(okHttpClient)
            .validateEagerly(true)
            .build()

    companion object {
        public const val NAME = "LastFMClient"
        public const val GSON_CONVERTER_FACTORY = "GSONConverterFactory"
        public const val RX_CALL_ADAPTER_FACTORY = "RXLastFMCallAdapterFactory"
    }

    override fun <T> api(service: Class<T>): T {
        return restAdapter.create(service)
    }
}
